### PR TITLE
Feature bucket delete improvement

### PR DESCRIPTION
**Pull request by dunedodo**

### CODE REVIEW

File: src/main/java/com/emc/object/s3/request/DeleteBucketRequest.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import com.emc.object.Method;
+import com.emc.object.util.RestUtil;
+
+import java.util.List;
+import java.util.Map;
+
+public class DeleteBucketRequest extends GenericBucketRequest {
+
+    // if this bucket contains objects/versions
+    private boolean hasBackgroundTasks;
```
Comment (twincitiesguy) [inline src/main/java/com/emc/object/s3/request/DeleteBucketRequest.java:12]: this field should be renamed and should follow the pattern of the other request objects as far as constructor arguments and getters/setters

File: src/main/java/com/emc/object/s3/S3Client.java [hunk 2]
```diff
@@ -110,6 +110,16 @@ public interface S3Client {
...
 
+    /**
+     * Deletes <code>bucketName</code>. The bucket could contain objects and versions before it can be deleted
+     */
+    void deleteBucket(DeleteBucketRequest request);
+
+    /**
+     * Gets status of background bucket cleanup tasks for <code>bucketName</code> if present
+     */
+    BucketDeletionStatus getBucketDeletionStatus(String bucketName);
```
Comment (xiaoxin-ren) [inline src/main/java/com/emc/object/s3/S3Client.java:121]: This API is tested for bucket with background deletion, what's the behavior for bucket not being deleted? Better add test case or more description.

File: src/main/java/com/emc/object/s3/jersey/S3JerseyClient.java [hunk 3]
```diff
@@ -332,6 +332,16 @@ public void deleteBucket(String bucketName) {
...
     }
 
+    @Override
+    public void deleteBucket(DeleteBucketRequest request) {
+        executeAndClose(client, request);
+    }
+
+    @Override
+    public BucketDeletionStatus getBucketDeletionStatus(String bucketName) {
+        return executeRequest(client, new GetEmptyBucketStatusRequest(bucketName), BucketDeletionStatus.class);
```
Comment (xiaoxin-ren) [inline src/main/java/com/emc/object/s3/jersey/S3JerseyClient.java:342]: Would suggest to change the request name to `GetBucketDeletionStatusRequest`.

File: src/main/java/com/emc/object/s3/bean/BucketDeletionStatus.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+@XmlRootElement(name = "EmptyBucketStatus")
+public class BucketDeletionStatus {
+    private String status;
+
+    private String created;
+
+    private String lastUpdated;
+
+    private long deleted;
```
Comment (xiaoxin-ren) [inline src/main/java/com/emc/object/s3/bean/BucketDeletionStatus.java:17]: Shall we keep the same as ECS to define as `Long entriesDeleted` ?
And also the rest of types to be class?

**Review by xiaoxin-ren — CHANGES_REQUESTED** [review]

File: src/main/java/com/emc/object/s3/request/DeleteBucketRequest.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import com.emc.object.Method;
+import com.emc.object.util.RestUtil;
+
+import java.util.List;
+import java.util.Map;
+
+public class DeleteBucketRequest extends GenericBucketRequest {
+
+    // if this bucket contains objects/versions
+    private boolean backgroundDeletion;
```
Comment (twincitiesguy) [inline src/main/java/com/emc/object/s3/request/DeleteBucketRequest.java:12]: Recommend changing the name here to something more meaningful - something like `recursivelyDeleteAllData`.

File: src/main/java/com/emc/object/s3/bean/BucketDeletionStatus.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+        this.lastUpdated = lastUpdated;
+        return this;
+    }
+
+    @XmlElement(name = "EntriesDeleted")
+    public Long getEntriesDeleted() {
+        return entriesDeleted;
+    }
+
+    public BucketDeletionStatus setEntriesDeleted(Long entriesDeleted) {
```
Comment (twincitiesguy) [inline src/main/java/com/emc/object/s3/bean/BucketDeletionStatus.java:94]: need javadocs for all getters/setters so users know what these fields mean.  you can copy descriptions from the API docs.

File: src/test/java/com/emc/object/s3/S3JerseyClientTest.java [hunk 7]
```diff
@@ -500,6 +500,89 @@ public void testDeleteBucketWithObjects() throws Exception {
...
+        Assert.assertEquals("EmptyBucketStatus should not be found", "NOT_FOUND", statusNotBeingDeleted.getStatus());
+        client.deleteBucket(new DeleteBucketRequest(bucketNameNotBeingDeleted, true));
+
+        // delete a bucket which is empty or not, the first status should always be PENDING.
+        client.deleteBucket(new DeleteBucketRequest(bucketName, true));
+        BucketDeletionStatus status = client.getBucketDeletionStatus(bucketName);
+        // check the status and dates of the bucket delete task, other entries are not given.
+        Assert.assertEquals("The initial status of a bucket which is being deleting should be PENDING.", status.getStatus(), "PENDING");
+        Assert.assertEquals("created date should be the same as lastUpdated date in EmptyBucketStatus.", status.getCreated(), status.getLastUpdated());
+
```
Comment (twincitiesguy) [inline src/test/java/com/emc/object/s3/S3JerseyClientTest.java:526]: would be good to somehow test the remaining status fields (counts, size, etc.)

File: src/test/java/com/emc/object/s3/S3JerseyClientTest.java [hunk 8]
```diff
@@ -500,6 +500,89 @@ public void testDeleteBucketWithObjects() throws Exception {
...
+        client.putObject(putObjectRequest);
+
+        client.deleteBucket(new DeleteBucketRequest(bucketName, true));
+        // Empty bucket execution frequency is 1 min, so 2 min would be enough.
+        // Wait until deletion background tasks are DONE.
+        Thread.sleep(2 * 60 * 1000);
+
+        BucketDeletionStatus status = client.getBucketDeletionStatus(bucketName);
+        // check the status and dates of the bucket delete task, other entries are not given.
+        Assert.assertEquals("The EmptyBucketStatus status of a bucket which is still in retention should be FAILED.", status.getStatus(), "FAILED");
```
Comment (twincitiesguy) [inline src/test/java/com/emc/object/s3/S3JerseyClientTest.java:582]: create multiple objects (i.e. include one without retention) and check other error fields to make sure they are populated (i.e. FailedToDeleteDueToRetention and EntriesDeleted)

**Review by twincitiesguy — CHANGES_REQUESTED** [review]

File: src/test/java/com/emc/object/s3/S3JerseyClientTest.java (hunk 7 above)
Comment (dunedodo) [reply src/test/java/com/emc/object/s3/S3JerseyClientTest.java:526]: still wait for dev to confirm if objectCount/totalSize are null.

File: src/main/java/com/emc/object/s3/bean/BucketDeletionStatus.java [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+     */
+    @XmlElement(name="Status")
+    public String getStatus() {
+        return status;
+    }
+
+    /**
+     * {@link BucketDeletionStatus#getStatus}
+     */
+    public BucketDeletionStatus setStatus(String status) {
```
Comment (twincitiesguy) [inline src/main/java/com/emc/object/s3/bean/BucketDeletionStatus.java:67]: this breaks the pattern of existing beans - setters should return `void` - all other bean classes have separate `with*` methods to use with chaining.  see CopyRange.java as an example.

File: src/main/java/com/emc/object/s3/request/DeleteBucketRequest.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        return recursivelyDeleteAllData;
+    }
+
+    /**
+     * {@link DeleteBucketRequest#isRecursivelyDeleteAllData}
+     */
+    public void setRecursivelyDeleteAllData(boolean recursivelyDeleteAllData) {
+        this.recursivelyDeleteAllData = recursivelyDeleteAllData;
+    }
+
```
Comment (twincitiesguy) [inline src/main/java/com/emc/object/s3/request/DeleteBucketRequest.java:50]: add `withRecursivelyDeleteAllData()` to use in chaining, similar to other request objects.

**Review by twincitiesguy — CHANGES_REQUESTED** [review]

**Review by twincitiesguy — APPROVED** [review]

**Review by xiaoxin-ren — APPROVED** [review]

**Pull request merged**